pkg/controlplane/peer/certs.go

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) The ClusterLink Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package peer
+
+import (
+	"errors"
+	"fmt"
+	"path"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
+
+	"github.com/clusterlink-net/clusterlink/pkg/util/tls"
+)
+
+// CertsConsumer represents a consumer of peer TLS certificates.
+type CertsConsumer interface {
+	SetPeerCertificates(parsedCertData *tls.ParsedCertData, rawCertData *tls.RawCertData) error
+}
+
+// parseError represents an error in parsing the peer certificates.
+type parseError struct {
+	err error
+}
+
+func (e parseError) Error() string {
+	return e.err.Error()
+}
+
+// CertsWatcher watches certificate updates.
+type CertsWatcher struct {
+	caPath   string
+	certPath string
+	keyPath  string
+
+	stopCh    chan struct{}
+	consumers []CertsConsumer
+
+	logger *logrus.Entry
+}
+
+// Name of the watcher.
+func (w *CertsWatcher) Name() string {
+	return "certs-watcher"
+}
+
+// AddConsumer adds a new peer certificates consumer.
+func (w *CertsWatcher) AddConsumer(consumer CertsConsumer) {
+	w.consumers = append(w.consumers, consumer)
```

Comment:
single goroutine access is assumed. Fine if it is internal, may want to protect (or at least document) if can be used by any part of the code.